packages/rpc-methods/src/restricted/getBip32PublicKey.ts

```diff
@@ -0,0 +1,169 @@
+   *
+   * @returns A promise that resolves once the extension is unlocked.
+   */
+  getUnlockPromise: (shouldShowUnlockRequest: boolean) => Promise<void>;
+};
+
+type GetBip32PublicKeySpecificationBuilderOptions = {
+  methodHooks: GetBip32PublicKeyMethodHooks;
+};
+
+type GetBip32PublicKeySpecification = ValidPermissionSpecification<{
+  permissionType: PermissionType.RestrictedMethod;
+  targetKey: typeof targetKey;
+  methodImplementation: ReturnType<typeof getBip32PublicKeyImplementation>;
+  allowedCaveats: Readonly<NonEmptyArray<string>> | null;
+  validator: PermissionValidatorConstraint;
+}>;
+
+type GetBip32PublicKeyParameters = {
+  path: ['m', ...(`${number}` | `${number}'`)[]];
+  curve: 'secp256k1' | 'ed25519';
+  compressed?: boolean;
+};
+
+/**
+ * The specification builder for the `snap_getBip32PublicKey` permission.
+ * `snap_getBip32PublicKey` lets the Snap retrieve public keys for a particular
+ * BIP-32 node.
+ *
+ * @param options - The specification builder options.
+ * @param options.methodHooks - The RPC method hooks needed by the method implementation.
+ * @returns The specification for the `snap_getBip32PublicKey` permission.
+ */
+const specificationBuilder: PermissionSpecificationBuilder<
+  PermissionType.RestrictedMethod,
+  GetBip32PublicKeySpecificationBuilderOptions,
+  GetBip32PublicKeySpecification
+> = ({ methodHooks }: GetBip32PublicKeySpecificationBuilderOptions) => {
+  return {
+    permissionType: PermissionType.RestrictedMethod,
+    targetKey,
+    allowedCaveats: [SnapCaveatType.PermittedDerivationPaths],
+    methodImplementation: getBip32PublicKeyImplementation(methodHooks),
+    validator: ({ caveats }) => {
+      if (
+        caveats?.length !== 1 ||
+        caveats[0].type !== SnapCaveatType.PermittedDerivationPaths
+      ) {
+        throw ethErrors.rpc.invalidParams({
+          message: `Expected a single "${SnapCaveatType.PermittedDerivationPaths}" caveat.`,
+        });
+      }
+    },
+  };
+};
+
+export const getBip32PublicKeyBuilder = Object.freeze({
+  targetKey,
+  specificationBuilder,
+  methodHooks: {
+    getMnemonic: true,
+    getUnlockPromise: true,
+  },
+} as const);
+
+export const getBip32PublicKeyCaveatSpecifications: Record<
+  SnapCaveatType.PermittedDerivationPaths,
+  CaveatSpecificationConstraint
+> = {
+  [SnapCaveatType.PermittedDerivationPaths]: Object.freeze({
+    type: SnapCaveatType.PermittedDerivationPaths,
+    decorator: (
+      method,
+      caveat: Caveat<
+        SnapCaveatType.PermittedDerivationPaths,
+        GetBip32PublicKeyParameters[]
+      >,
+    ) => {
+      return async (args) => {
+        const { params } = args;
+        validatePath(params);
+
+        const path = caveat.value.find(
+          (caveatPath) =>
+            isEqual(params.path, caveatPath.path) &&
+            caveatPath.curve === params.curve,
+        );
+
+        if (!path) {
+          throw ethErrors.rpc.invalidParams({
```

Comment:
This one should be `ethErrors.rpc.unauthorized`.

Reply:
Same goes for the other implementations then I guess?

Reply:
Well, it depends. In this case, the parameters are valid, meaning that they are well-formed and we understand them. The problem here is that the parameters are incompatible with the attenuation implemented by the caveat, which means that the invoker (i.e. the caller of the RPC method) does not have the requisite authority to do what they're trying to do, which means that we should use `unauthorized`. So, any failures of that nature should always throw that other.

And just for the sake of being explicit, the use of `invalidParams` in the permission specification builder in this PR is a good example of an appropriate use of that error.

Reply:
> So, any failures of that nature (call them "permission failures") should always throw that error.
That is what I meant as well, I imagine we have a similar implementation for the other key related RPC calls.

Reply:
https://github.com/MetaMask/snaps-skunkworks/blob/main/packages/rpc-methods/src/restricted/getBip44Entropy.ts#L188-L191

https://github.com/MetaMask/snaps-skunkworks/blob/main/packages/rpc-methods/src/restricted/getBip32Entropy.ts#L234-L237